Feature/persistence

serpient opened the pull request:
![gif](https://media.giphy.com/media/loA2abnhxZU7DTHp8D/giphy.gif)
# Things Done
### Game Data Serializer + Deserializer
- takes in game input and compiles the data fields, converts to JSON
- takes in JSON and deserialize back to valid game data
- had to adjust the Lite 3 struct for positions - made it public so it can be used in the deserializer

### Repository
- repository class with 2 duck types -> Local (postgres) and Memory

### ReInitializer
- recreates the game instance

### Persister class
- puts together all of the above to suspend and resume a game. 

### GameCreator class
- handles the creation of a new game instance based on whether its a new or suspended game per user input.

serpient committed:
game serializer / deserializer

- takes in game input and compiles the data fields, converts to JSON
- takes in JSON and deserialize back to valid game data
- had to adjust the Lite 3 struct for positions ; make it public so it can be used in the deserializer

serpient committed:
starter working connection to local postgres db

ssunday commented on lib/persister/game_data.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def status
+        {
+            :name => :status,
```
`name: :status,`

ssunday replied:
The different ruby syntax

serpient replied:
@ssunday to clarify, are you saying its more common to write hashes like `name: :status` over `:name => :status`

ssunday replied:
I prefer it, it's a lot cleaner. Rubocop also suggests to use it

ssunday commented on spec/game_serializer_spec.rb:
```diff
@@ -0,0 +1,61 @@
+    attr_accessor :game, :lite_3_mode, :deserialized_output, :game_serializer, :seralized_output
+
+    before(:each) do
+        @game = Game.new(
```
Could you use `let` instead of @game, etc?

ssunday commented on spec/game_serializer_spec.rb:
```diff
@@ -0,0 +1,61 @@
+                {"token":"X","position":5}],
+            "player_2_moves":[{"token":"O","position":3}]
+        }
+        @deserialized_output = {
```
There's a LOT of logic here

serpient committed:
adjust serialization test cases

- extract repeated items
- add regular mode test case
- add lite 3, no moves, test case

serpient committed:
change hash syntax to `name:` to match best practice

serpient committed:
add types to ducktypes

- easier to reinitialize when each class knows its corresponding type

serpient committed:
extract mock test helper fns into separate file

serpient committed:
fix some strange merge issues

serpient committed:
initializer outputs a new Game with existing data injected

- clean up test units
- add additional helper fns for testing
- delete duplicate game serializer test file
- rename module to reinitializer. public method is generate

serpient committed:
change hash access to fetch

serpient committed:
repository interface with memory and local db types

- inject the storage type and table name

ssunday commented on lib/repository/local.rb:
```diff
@@ -0,0 +1,56 @@
+    attr_accessor :db, :table
+
+    def initialize(table_name)
+        @db = Sequel.connect('postgres://postgres:postgres@localhost/postgres')
```
could the host connection be a constant?

serpient committed:
remove starter database file

serpient committed:
Persister class to save/resume a game

- takes in repository_type (so far, memory and local db)
- suspends the game and returns the game_id of last saved
- resumes takes in game_id and recreates the game instance

serpient committed:
retrieve_list of saved game ids

ssunday commented on lib/repository/local.rb:
```diff
@@ -41,6 +47,12 @@ def retrieve_all_fields
     end
 
+    def retrieve_id_and_date
+        sql = "SELECT id, created_at from #{table} ORDER BY created_at DESC LIMIT 6"
```
why 6? What does 6 represent?  Is it to get the last 3 turns of each player?

serpient replied:
I'm using 6 arbitrarily to limit the amount of suspended games I retrieve

jayden replied:
It may make sense to extract this as a constant to describe this

serpient committed:
First pass of Persister Integration in TTT

- saved game vs new game in tic tac toe. Will need to move this out into its own class -- maybe called game_start
- add save outputs and validation
- add save game function within game
- add ability to inject setting types into create_settings function. so it can be used twice -- first for paused vs new game, then the rest of the settings for a new game
- add resume_game setting type
- add a disconnect fn to repository -- fixes bug where I had opened too many connections. need to disconnect after I run my tests

serpient committed:
new game creator class

- use options hash for Sequel connect to make it clearer what the params are
- extract game creation in tic tac toe into its own class called GameCreator. handles the logic of suspended or new game instances
- cleaned up game setting setter to only have 1 public method again.

serpient committed:
extract save input to InputType

serpient committed:
game_data decoupled from game to output types

- change GameData to module, and remove dependency on game as a constant to grab deserialized  functions. Only use game to grab data for serializer_types.
- game serializer changed to module as well, as it doesn't need to hold any state.
- update persister to use module GameSerializer
- add gameData spec and update other tests to use module instead of class

serpient committed:
use fetch for hash retrieval

ssunday commented on lib/persister/game_data.rb:
```diff
@@ -1,10 +1,26 @@
+
+    def serializer_types(game)
         @game = game
+        types.map do |type|
```
Could you use `yield` to reduce duping in deleting certain keys from the types

serpient replied:
Ended up using a reduce fn instead to pick out my desired keys. Thought it would be clearer than trying to delete a key, and its immutable

```
    def serializer_types(game)
        @game = game
        types.reduce([]) do |array, type|
            array.push({ 
                    name: type.fetch(:name), 
                    data: type.fetch(:data) 
                })
            array
        end
    end

    def deserializer_types
        types.reduce([]) do |array, type|
            if type.fetch(:deserialize, false)
                array.push({ 
                    name: type.fetch(:name), 
                    deserialize: type.fetch(:deserialize) 
                })
            end
            array
        end
    end
```

ssunday commented on lib/persister/game_data.rb:
```diff
@@ -19,82 +35,77 @@ def types
         {
             name: :status,
-            data: game.status,
+            data: game && game.status,
```
game&.status http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/

serpient replied:
would this: `data: game&.game_state&.board&.positions` still be better than `data: game && game.game_state.board&.positions` considering they would all need to change? Just curious --

ssunday replied:
safe navigator is easier to read and cleaner than doing &&. If I saw that in a code-base, I'd correct it all to use the &

ssunday replied:
I prefer chaining the safe navigator operator than doing that hard && check when possible.

serpient replied:
gotcha. Will update

jayden commented on lib/repository/memory.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def disconnect
+        initialize_table
```
The added indirection seems unnecessary here since we only call `initialize_table` in here and it only called once in this file. Inlining `@db = {}` seems fine to me.

jayden commented on lib/repository/memory.rb:
```diff
@@ -0,0 +1,40 @@
+            created_at: Time.new(2010,8,10,db.fetch(table).length),
+            id: rand(1000)
+        })
+        nil
```
Why are we explicitly returning a `nil` here?

jayden commented on lib/game_setting/game_settings.rb:
```diff
@@ -17,6 +19,15 @@ def types
+            :name => SettingTypes::NEW_OR_SUSPENDED_GAME,
+            :message => Messages.resume_game_options(list_of_saved_games),
+            :valid? => -> (input: nil) { Validator.resume_option_valid?(input: input, valid_game_ids: valid_game_ids) },
+            :parse => -> (input:) { input == "N" ? :new_game : input.to_i },
```
Should this be declared in the `InputType` module?

jayden commented on lib/game.rb:
```diff
@@ -36,31 +38,45 @@ def initialize(
+        game_io.output_message(Messages.output_save_option)
+    end
+
+    def get_valid_input(board:, presenter:)
```
Bit of a nitpick, but it is typically conventional to avoid prefixing method names with `get` or `set` for accessors and mutators. (Reference: https://github.com/rubocop-hq/ruby-style-guide#accessor_mutator_method_names)

jayden commented on lib/game.rb:
```diff
@@ -78,5 +94,14 @@ def output_win
     end
+
+    def save_game
+        repository_type = game_io.is_a?(ConsoleIO) ? Memory.new : Local.new
```
Why do we need to inspect the class of `game_io` to determine which type of repository to use?

serpient replied:
Didn't think this through when I added -- adjusted to use dependency injection into Game itself, with a default of Local.new

jayden commented on Gemfile:
```diff
@@ -5,4 +5,6 @@ group :development do
   gem 'guard-rspec'
   gem 'pry'
+  gem 'sequel'
+  gem 'pg'
```
Does the README need to be updated with additional steps so that we can setup our local environment to run this app?

jayden commented on lib/game_creator.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    def is_a_suspended_game_id?(input)
+        input.is_a?(Integer)
```
Seems strange to me that any integer input would satisfy this since it sounds like any number would be a valid suspended game id.

serpient replied:
so there's already a validation check when grabbing user input -- this check is whether the input is a integer, in which case the user is asking for a suspended game, OR, "N" (string) for a New Game.

the validation check occurs during create_settings
```
 def new_or_suspended_game_input
        suspended_game_ids = persister.suspended_game_ids(suspended_games)
        new_or_suspended_game_settings = GameSettings.new_or_suspended_game(suspended_games, suspended_game_ids)
        setting_output = game_settings.create_settings(settings: new_or_suspended_game_settings)
        setting_output.fetch(SettingTypes::NEW_OR_SUSPENDED_GAME)
    end
```

serpient replied:
the naming is confusing cause it seems its a validation.
maybe pull this into validation file, so its grouped with the others.

serpient committed:
convert to safe navigation operator &.

jayden commented on lib/persister/game_data.rb:
```diff
@@ -0,0 +1,130 @@
+    def game_mode
+        {
+            name: :game_mode,
+            data: game && game.game_state.game_mode.type,
```
If you find yourself reaching really far into an object to retrieve some value, it is often considered a code smell. In this case, game seems like some kind of god class that knows way too much about the internals of its subcomponents. 

Refer: https://en.wikipedia.org/wiki/Law_of_Demeter

serpient replied:
I did think I was toeing the line with how far I was reaching -- but I couldn't think of another way of accessing all of this info other than to have getter fns that stores the data.

for example, game_state would then keep track of the game_mode type, etc.

this was something I'd like to discuss more with you guys during code review :D

aaizuss replied:
This might also be helpful reading https://dzone.com/articles/the-genius-of-the-law-of-demeter

serpient committed:
refactor

- remove unnecessary indirection in memory.disconnect
- remove unused nil in memory.save
- use InputType::NEW_GAME , instead of a hardcoded "N"
- remove use of "get" in function names
- remove repository check in save_game. use dependency injection into game itself for repository type, with default of local
- update README with new postgres setup instructions
- adjust local database table creation to create a new table if it doesn't exist

serpient committed:
adjust deserializer and serializer data types

- pick out the wanted data instead of deleting unwanted keys
- this streamlines it within deserializer as well, as it no longer needs to check for whether it has a deserializer. the entire array has data types only with deserializer fns

jayden commented on README.md:
```diff
@@ -14,3 +14,11 @@
+# Setting up Local PostgreSQL Database 
+1. Install PostgreSQL if you dont have it. For mac users, try `brew install postgresql`
+2. Start your postgres instance -- `brew services start postgresql`
+3. Set up user. 
```
It may be useful here to provide more detailed instructions on how to create a user and database. Put yourself in the shoes of someone who would be setting up this project for the first time and may not have intimate knowledge of Postgres. We should aim to make the setup as painless as possible by providing a good README.

It's also worth noting that postgres needs to be installed prior to installing the dependencies with `bundle install` as the `pg` gem would fail to install without it.

jayden commented on lib/game.rb:
```diff
@@ -31,36 +34,51 @@ def initialize(
+    def input_not_valid(input:, board:) 
+        !Validator.position_valid?(input: input, board: board) && !Validator.save_game?(input: input)
+    end
+
```
I can't comment within `update_game_status` since it's not part of the diff here, but I wanted to point out that both conditionals would be evaluated even if the first one was true. We may want to convert it to an `if/elsif` to avoid this.

jayden commented on lib/game.rb:
```diff
@@ -31,36 +34,51 @@ def initialize(
         end
         output_board
+        save_game if status == :save
         output_win if status == :win
```
Similar comment here about the branching logic - may make sense to change this to an `if/elsif` since it would never be both a win and a tie.

serpient replied:
would if / else be better than just doing an explicit return?
```
return save_game if status == :save
return output_win if status == :win
return output_tie if status == :tie 
```

jayden reviewed (commented):
<img width="817" alt="Screen Shot 2019-05-24 at 12 31 38 PM" src="https://user-images.githubusercontent.com/649702/58354984-d7271d80-7e27-11e9-847a-ec6007954b04.png">

I get this error when I try to run the game. Anyone else seeing this?

serpient commented:
@jayden i'm updating the README now 👍 
also, i mistyped the executable instructions, should be:
`ruby bin/tic_tac_toe.rb`

serpient committed:
update README with further set up instructions

serpient committed:
refactor: remove game_state

- lift data from sub classes and access directly from game class
- remove game_state and tests. Move all logic into game

serpient committed:
use doubles to test GameCreator

- inject Persister class into GameCreator instea of instantiating within the initializer. this makes it easier to inject a double during testing
- use doubles to control the returned messages of the numerous methods used for the create class
- tested 3 cases: new game with no suspended games in history, new game choice even with the option for suspended games, and resuming a suspended game

serpient committed:
remove comments
- feedback about SQL: use columns and relational database strengths next time instead of JSON blob.

jayden reviewed (approved):
Not dealbreakers, but I just wanted to point out that you were declaring attr_accessors in your tests when they aren't necessary. Another thing that I noticed is that there are multiple parent describe blocks in your tests. It's preferable to structure it so that there is a single parent describe block with nested describe/context blocks as it makes the test output easier to read. Example:

```ruby
describe Reinitializer
  context "Lite 3 Mode - No Moves" do
    describe "generate" do
      it "does something" do
      end
    end

  context "Lite 3 Mode - With Moves" do
    describe "generate" do
      it "does something" do
      end
    end
  end

  context "Regular Mode - Computer Player 2 - Board Size 4" do
    describe "generate" do
      it "does something" do
      end
    end
  end
end
```
generates a similar output:
<img width="307" alt="Screen Shot 2019-05-31 at 3 32 49 PM" src="https://user-images.githubusercontent.com/649702/58738142-62b92500-83b9-11e9-8e4d-b7543c0fbdab.png">
